simplify_work/simplify_algorithms.py

```diff
@@ -0,0 +1,305 @@
+"""
+Python versions of the algorithms from the paper.
+"""
+from __future__ import print_function
+from __future__ import division
+
+import sys
+import random
+import tempfile
+import argparse
+import heapq
+import math
+
+import msprime
+import numpy as np
+
+from six.moves import StringIO
+
+from algorithms import *
+
+class SearchablePopulation(Population):
+    '''
+    Like Population, but allows searching ancestors by index.
+    '''
+    def from_index(self, index):
+        # obviously should do this differently
+        for x in self._ancestors:
+            if x.index ==index:
+                return x
+        return None
+
+class Simplifier(Simulator):
+    """
+    Modified from Simulator().
+    """
+    def __init__(self, ts, samples, max_segments=100):
+        # since we will have no migration events,
+        # need to use one population but copy over population information
+        N = 1
+        sample_size = len(ts.samples())
+        num_loci = ts.sequence_length
+
+        self.ts = ts
+        self.n = sample_size
+        self.m = num_loci
+        # self.r = recombination_rate
+        # self.migration_matrix = migration_matrix
+        self.max_segments = max_segments
+        self.segment_stack = []
+        self.segments = [None for j in range(self.max_segments + 1)]
+        for j in range(self.max_segments):
+            s = Segment(j + 1)
+            self.segments[j + 1] = s
+            self.segment_stack.append(s)
+        self.P = [SearchablePopulation(id_) for id_ in range(N)]
+        self.C = []
+        self.L = FenwickTree(self.max_segments)
+        self.S = bintrees.AVLTree()
```

Comment:
I don't think we need S here. This is used to track the amount of ancestral material present in each segment, so that we know when Hudson's algorithm is complete.